### PR TITLE
Tgui stat panel

### PR DESCRIPTION
**Pull request by PowerfulBacon**
**Fix chat HARDER ports:**
 - https://github.com/tgstation/tgstation/pull/54030/files
 - https://github.com/tgstation/tgstation/pull/54517

## Important
client.verbs +=, client.verbs -= and client.verbs |= is now not allowed and will not function properly. Once this is merged, all PRs will be required to do client.add_verb()

## About The Pull Request

Adds in the beautiful TGUI stat panel.
Actually includes all of the verbs.

Completely depreciates statclick and /datum/verbs

## Why It's Good For The Game

It looks good.
MC no longer lags the server.

**NOTE: TABS WERE NOT SORTED ALPHABETICALLY IN THESE SCREENSHOTS (Added that afterwards)** (actually I haven't fixed that yet even 1 day later)  <-- I have now

Status:
image
image
image
image
(Not lagging the game or the server, and updating in real time)
image
image
image
image

## Todo:
 - [x] Make it so you can resize the stat panel.
 - [x] Clean up some ungodly stat code.
 - [x] Fix the single ticket manager issue

## Important things to note / Potential Issues:
This completely removes the stat panel, so if TG chat fails to load the only way to use the fix-chat verb (which barely ever fixes chat) is to type in the command line. I'll port something that adds a message to chat from TG to help with this, but for now we have a message instead.

## Changelog
:cl:
add: TGUI stat panel
del: Old stat panel
del: Depreciates /datum/verbs
del: Depreciates statclick
refactor: Refactored how verbs are added to things
refactor: Refactors stat code
fix: Ports fix fix chat from TG station
/:cl:

### CODE REVIEW

**Comment by Moccha-Bee** [comment]
papa bless

**Comment by PowerfulBacon** [comment]
Ready

**Comment by Archanial** [comment]
okay but how does it look with proposed 17x15?

**Comment by PowerfulBacon** [comment]
> 
> 
> okay but how does it look with proposed 17x15?

Should look pretty similar, the tab selection might scroll though but that can be changed to wrap to new line if needed. Also I think I forgot to update the alphabetical order bars

**Comment by ike709** [comment]
image

Status panel's values have no spacing

**Comment by ike709** [comment]
IC notes seem to not work at all.

**Comment by ike709** [comment]
CPU usage is maxing out and I think it's this PR's fault.

**Comment by ike709** [comment]
A bunch of verbs are supposedly missing.

**Comment by ike709** [comment]
Several players report layout issues but I have no screencaps.

**Comment by ike709** [comment]
Confirmed this was the cause of CPU issues.

**Comment by PowerfulBacon** [comment]
TODO:
 - Put sending messages into middleware.
 - Stat panel subsystem

**Comment by PowerfulBacon** [comment]
Recommended with https://github.com/BeeStation/BeeStation-Hornet/pull/2668

**Comment by PowerfulBacon** [comment]
Player panel shows under admin for non admins but cannot be used.
Chat seems to be cut off for almost all players (Chat is slightly too big for non native resolution users).
There is still no gab between the text stats
https://cdn.discordapp.com/attachments/427337870844362755/770600184614617098/Screenshot_1.png
Gas scan is missing
Debug admin log on finding ticket with ID [id]

**Comment by ike709** [comment]
Conflicts.

**Comment by Cenrus** [comment]
Observe, Toggle HUD, check health, and gas scan verbs are missing from ghosts

**Comment by PowerfulBacon** [comment]
On it

**Comment by IllotAFV** [comment]
OOC is missing Ghost tab.

**Comment by Wilchenx** [comment]
image
Normal players can see player panel which should not be visible, button itself doesn't do anything without perms.

**Comment by PowerfulBacon** [comment]
Huh that's an odd one

**Comment by PowerfulBacon** [comment]
Seemed to cause a lot of lag too, might have been unrelated but highly unlikely.

**Comment by ike709** [comment]
How's this going? Also, conflicts.

**Comment by PowerfulBacon** [comment]
Going through some fixes now, fixing the data lag *might* be a little more hard

**Comment by PowerfulBacon** [comment]
Fixed all cases of missing verbs

**Comment by PowerfulBacon** [comment]
Fixed player panel bug

**Comment by PowerfulBacon** [comment]
Last thing to do is move the transfer of verb data to the verb on gain rather than on switch tab to remove data lag.

**Comment by ike709** [comment]
Conflicts

**Comment by PowerfulBacon** [comment]
Data lag should be slightly less, large tabs will now update slower and the stat subsystem is background

**Comment by PowerfulBacon** [comment]
useLocalState(context, 'number', settings.statSize); can be null?

**Comment by PowerfulBacon** [comment]
Object verbs were lost

**Comment by IllotAFV** [comment]
OOC is missing a Ghost tab, OOC also doesnt have Observe tab.

**Comment by ike709** [comment]
Conflicts

**Comment by PowerfulBacon** [comment]
ok this is a bit boring but going to try and get it done today. Should be easy to grab the verbs of things the player is wearing since I just realized its mob.contents

**Comment by PowerfulBacon** [comment]
Changelog:
 - Fixed object verbs not appearing
 - Added options to change between scrolling tabs and tabs wrapping to the new line
 - Sends only the tab data rather than literally all the verbs 4noraisin when we don't need to send to improve data lag.

Final Todo:
 - Fix OOC and Object mob verbs not showing up
 - Test for chat crashes

**Comment by PowerfulBacon** [comment]
GetFlatIcon is just way too laggy to be useful

File: code/controllers/subsystem/stat.dm [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	//Mobs are more important than items.
+	//Mob icons will change if their name changes, their type changes or their overlays change.
+	if(istype(A, /mob))
+		var/mob/M = A
+		var/overlay_hash = ""
+		for(var/image/I as() in M.overlays)
+			overlay_hash = "[overlay_hash][I.icon_state]"
+		what_to_search = "[M.type][M.name][overlay_hash]"
+	//Makes it shorter
+	what_to_search = sha1(what_to_search)
```
Comment (ike709) [inline code/controllers/subsystem/stat.dm]:
```suggestion
	what_to_search = md5(what_to_search)
```

My bad.

**Comment by PowerfulBacon** [comment]
GetFlatIcon has been replaced. There is now no spike in the subsystem when grabbing icons. 🎉

File: code/controllers/subsystem/stat.dm [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+				M.UpdateMobStat(FALSE)
+
+		if (MC_TICK_CHECK)
+			return
+
+//Note: Doesn't account for decals on items.
+//Whoever examins an item with a decal first, everyone else will see that items decals.
+//Significantly reduces server lag though, like MASSIVELY!
+/datum/controller/subsystem/stat/proc/get_flat_icon(atom/A)
+	var/what_to_search = "[A.type][A.icon_state[1]]"
```
Comment (park66665) [inline code/controllers/subsystem/stat.dm]:
```suggestion
	var/what_to_search = "[A.type][A.icon_state ? A.icon_state[1] : "*"]"
```

It turns out, tables, full tile windows and shits do have an empty string as `icon_state`. The whole proc runtimes and return when it tries to access the first letter of an empty string.

File: code/controllers/subsystem/stat.dm [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	//Mobs are more important than items.
+	//Mob icons will change if their name changes, their type changes or their overlays change.
+	if(istype(A, /mob))
+		var/mob/M = A
+		var/overlay_hash = ""
+		for(var/image/I as() in M.overlays)
+			if(istext(I.icon_state))
```
Comment (park66665) [inline code/controllers/subsystem/stat.dm]:
```suggestion
			if(I.icon_state && istext(I.icon_state))
```

Gotta check here if `icon_state` is empty here too

File: code/controllers/subsystem/stat.dm (hunk 2 above)
Comment (PowerfulBacon) [reply code/controllers/subsystem/stat.dm]: Yikes

File: code/controllers/subsystem/stat.dm (hunk 3 above)
Comment (PowerfulBacon) [reply code/controllers/subsystem/stat.dm]: Oh I thought istext catches null

File: tgui/packages/tgui-panel/stat/StatText.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+      <Button
+        content={title}
+        onClick={() => sendMessage({
+          type: 'stat/pressed',
+          payload: {
+            action_id: action_id,
+            params: params,
+          },
+        })}
+        color="transparent"
```
Comment (PowerfulBacon) [inline tgui/packages/tgui-panel/stat/StatText.js:181]: this should take into account the button colour option in settings.

**Comment by ike709** [comment]
YOLO

**Comment by francinum** [comment]
Why the fuck did you merge it in this state, what the fuck are you on?

**Comment by VictorVow** [comment]
[RIP Linux](https://github.com/BeeStation/BeeStation-Hornet/issues/3363)

**Pull request merged**